lib/Dialect/TritonGPU/Transforms/Coalesce.cpp

```diff
@@ -19,86 +22,159 @@ using namespace mlir::triton;
+      struct Layout {
+        unsigned maxPerThread = 0;
+        unsigned minPerThread = 0;
+        unsigned elemNumBytes = 0;
+      };
 
-    unsigned perThread = getNumElementsPerThread(op, order, axisInfoAnalysis);
-    LDBG("perThread for op: " << perThread);
+      unsigned sharedPerThread = 0;
+      llvm::SmallVector<Layout, 10> layouts;
+      for (auto MI = slices.member_begin(I); MI != slices.member_end(); ++MI) {
+        auto op = *MI;
+        Value ptr = getMemAccessPtr(op);
+        auto tensorType = ptr.getType().cast<RankedTensorType>();
+        auto shapePerCTA = triton::gpu::getShapePerCTA(tensorType);
+        auto contiguity = axisInfoAnalysis.getAxisInfo(ptr)->getContiguity();
+        SmallVector<unsigned> order = argSort(contiguity);
+        int numElems = product<int64_t>(shapePerCTA);
+        int numThreads = numWarps * threadsPerWarp;
+        unsigned elemNumBits = getElementBitWidth(tensorType);
+        unsigned elemNumBytes = std::max(elemNumBits / 8, 1u);
 
-    for (Operation *opSameOrder : memAccessesSameOrder) {
-      if (opSameOrder == op)
-        continue;
-      unsigned currPerThread =
-          getNumElementsPerThread(opSameOrder, order, axisInfoAnalysis);
-      LDBG("perThread for opSameOrder: " << currPerThread);
-      perThread = std::max(perThread, currPerThread);
-    }
+        // maximum number of consecutive elements to avoid breaking cross-thread
+        // memory coalesing.
+        unsigned maxMemopSize = 16;
+        unsigned perThread =
+            getNumElementsPerThread(op, order, axisInfoAnalysis);
+        unsigned maxPerThread =
+            std::max(perThread, maxMemopSize / elemNumBytes);
 
-    perThread = std::min<int>(perThread, std::max(numElems / numThreads, 1));
-    LDBG("perThread: " << perThread);
-
-    if (!dyn_cast<triton::LoadOp>(op)) {
-      // For ops that can result in a global memory write, we should enforce
-      // that each thread handles at most 128 bits, which is the widest
-      // available vectorized store op; otherwise, the store will have "gaps"
-      // in the memory write at the warp level, resulting in worse performance.
-      // For loads, we can expect that the gaps won't matter due to the L1
-      // cache.
-      unsigned elemNumBits = getElementBitWidth(refTensorType);
-      perThread = std::min<int>(
-          perThread, getNumElementsPerThread(op, order, axisInfoAnalysis));
-    }
-    SmallVector<unsigned> sizePerThread(refTensorType.getRank(), 1);
-    sizePerThread[order[0]] = perThread;
+        // minimum number of consecutive elements to ensure a saturation of one
+        // memory transaction.
+        unsigned memTranscationSize = 128;
+        unsigned minPerThread =
+            memTranscationSize / (threadsPerWarp * elemNumBytes);
+        // minimum number of consecutive elements to ensure a four-byte
+        // alignment.
+        minPerThread = std::max(minPerThread, 4 / elemNumBytes);
```

Comment:
what happens for 64bits type?

Reply:
This is mainly for sub byte types. For f64, alignment isn't going to be an issue? i.e, any value of `minPerThread` should just be fine from alignment point of view?